presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Key.java

```diff
@@ -0,0 +1,159 @@
+/*
```

Comment:
I think this class and next one belong in spi. These are not specific to planner per se and can be useful in other places.

Reply:
> I think this class and next one belong in spi. These are not specific to planner per se and can be useful in other places.

I see these as internal optimizer data structures. They are derived initially from reading catalog constraints during from examining catalog metadata versus via optimizer interaction with the connector API. Please provide a motivating example of where you would see the KeyProperty, MaxCard, etc. being used outside of the optimizer. Happy to have that discussion. Thanks for your comment.